src/libsystemd-network/sd-dhcp-client.c

```diff
@@ -417,7 +417,7 @@ int sd_dhcp_client_set_hostname(
 
         /* Refuse hostnames that neither qualify as DNS nor as Linux hosntames */
         if (hostname &&
-            !(hostname_is_valid(hostname, false) || dns_name_is_valid(hostname) > 0))
+            !(hostname_is_valid(hostname, false) && dns_name_is_valid(hostname) > 0))
```

Comment:
this change would mean a name has to both qualify as valid linux hostname and as DNS hostname. That wasn't the intention here. We wanted to be fine with either, as the comment suggests...

Reply:
That is what I intended since only that rejected invalid labels such as the one I added in the test:

    assert_se(sd_dhcp_client_set_hostname(client, "bla.*") == -EINVAL);

But I agree, the comment suggests otherwise. It would need update too.

What is correct anyway? It seems that dns_name_is_valid does not reject illegal characters... The comment at hostname_is_valid seems to indicate that hostname_is_valid is a subset of dns_name_is_valid. Should we use hostname_is_valid only?

Reply:
both functions validate according to different rules. some restrictions are made by linux hostnames that DNS doesn't make (for example, linux insists on a max length of 63 chars, while DNS allows 255), and others are made by DNS that Linux doesn't make (such as no empty labels and stuff).

It's not entirely clearly what rules the DHCP provided hostname has to follow, we hence are lenient here, and permit all names that validate with at least one of the two schemes.

(Though one thing to mention: our interpretation of linux hostnames is more restrictive than what the kernel actually allows even, because the kernel for example allows empty hostnames and all that stuff, and really don't want to allow that).

Reply:
Ok I see, with that in mind following what the comments says make sense.

However, then maybe dns_name_is_valid is a bit too loose? Allowing characters such as star seems wrong to me. Maybe it should check characters to be according to 
https://tools.ietf.org/html/rfc1035#section-2.3.1

Reply:
I realized that dns_name_is_valid succeeds with something like "test.*" because it escapes the DNS string and then verifies it. However, the DHCP code uses dns_name_to_wire_format which does not escape strings and illegal characters hit the wire... This is still not a problem in practise since callers of sd_dhcp_client_set_hostname validate strings provided by users on their side too (config_parse_hostname).